### ISSUE TITLE
hwloc v2.9.2 default type filtering does not seem to match documentation

### ISSUE THREAD

**Issue by HadrienG2** [open-mpi/hwloc#622]
The documentation of `hwloc_type_filter_e` for hwloc v2.9 [states the following](https://hwloc.readthedocs.io/en/v2.9/group__hwlocality__configuration.html#ga9a5a1f0140cd1952544477833733195b):

> By default, most objects are kept ([HWLOC_TYPE_FILTER_KEEP_ALL](https://hwloc.readthedocs.io/en/v2.9/group__hwlocality__configuration.html#gga9a5a1f0140cd1952544477833733195bafda7b59e6810dfe778d8f9a4cc1e350e)). Instruction caches, I/O and Misc objects are ignored by default ([HWLOC_TYPE_FILTER_KEEP_NONE](https://hwloc.readthedocs.io/en/v2.9/group__hwlocality__configuration.html#gga9a5a1f0140cd1952544477833733195ba4f835955414de92c77d99b8419d4647a)). Die and Group levels are ignored unless they bring structure ([HWLOC_TYPE_FILTER_KEEP_STRUCTURE](https://hwloc.readthedocs.io/en/v2.9/group__hwlocality__configuration.html#gga9a5a1f0140cd1952544477833733195ba7664716643bf1db83e631eed34f659e4)).

However, on my system, if I build a topology with the default configuration (`hwloc_topology_init()` immediately followed by `hwloc_topology_load()`) and probe every object type using `hwloc_topology_get_type_filter()`, I observe the following two discrepancies:

- Dies have a default filtering policy of HWLOC_TYPE_FILTER_KEEP_ALL where the docs claim it should be HWLOC_TYPE_FILTER_KEEP_STRUCTURE
- MemCaches have a default filtering policy of HWLOC_TYPE_FILTER_KEEP_NONE while the docs do not explicitly mention them, suggesting it should be HWLOC_TYPE_FILTER_KEEP_ALL

Another thing I noticed along the way is that calling `hwloc_topology_get_type_filter()` on a topology which has not been built yet does not seem supported (it leads to various nondeterministic `hwloc_topology_check()` assertion errors down the line), which I found a bit surprising as the structure of the documentation suggests that `hwloc_topology_get_type_filter()` would be one of those few functions that are safe to call before loading the topology.

**Comment by bgoglin** [comment]
Hello, Indeed there was some discussion of no showing Dies when there's a single Die per Package, but it wasn't clear if Linux would do it, or if the hwloc Linux backend would, or if it should be done in these generic filters. I guess the doc was forgotten in some updates.

MemCaches is indeed "none" just like instruction caches.

Filters are initialized during init(), it should work fine before load() then. But I am not sure what check() is supposed to do before load().

Thanks, I'll have a look at all this.

**Comment by HadrienG2** [comment]
Thanks, after cross-checking the issue indeed only happens if I call check() before load(). I removed the offending assertion since that's obviously not something I should do :)

For the record, it results in this sort of assertion failure:

```text
topology.c:5077: hwloc_topology_check: Assertion `!topology->modified' failed.
```

This is with libhwloc v2.9.2 as packaged by openSUSE Tumbleweed (distro version 2.9.2-1.2).

**Comment by bgoglin** [comment]
check() is documented as after load() only anyway. I improved it to return before checking what cannot be checked before load().

Filter doc was updated as well.

Dies were confusing because recent Linux kernels always expose die info even when the hardware doesn't support dies. Hence we'd always get a die, and we cannot know if it's really a single die in a recent CPU or a fake die in a old CPU. Therefore our Linux backend will drop those useless and often-non-existing dies regardless of how die filtering was configured in the core.
